test/distributed/tensor/parallel/test_tp_random_state.py

```diff
@@ -0,0 +1,185 @@
+        #   2. residual dropout which belongs to a non-tensor-parallel region
+        mesh = torch.arange(self.world_size).reshape(2, 2)
+        device_mesh = DeviceMesh(self.device_type, mesh)
+        tp_rank = device_mesh.get_coordinate()[0]  # the tensor parallel dimension is 0
+        dp_rank = device_mesh.get_coordinate()[1]  # the data parallel dimension is 1
+
+        attn_dropout = torch.nn.Dropout(p=0.2)
+        attn_dropout_tp = parallelize_module(
+            attn_dropout,
+            device_mesh,
+            PairwiseParallel(make_input_shard_1d_last_dim, make_sharded_output_tensor),
+        )
+        resid_dropout = torch.nn.Dropout(p=0.2)
+        resid_dropout_tp = parallelize_module(
+            resid_dropout,
+            device_mesh,
+            PairwiseParallel(make_input_replicate_1d, make_output_tensor),
+        )
+
+        for enable_distribute_flag in [False, True]:
+            # disable/enable parallel RNG feature
+            random._rng_tracker.distribute_region_enabled = enable_distribute_flag
+
+            inp = torch.ones(*local_shape, device=self.device_type)
+            out = attn_dropout_tp(inp)
+            # examine the output
+            # check within TP groups
+            # all-gather local shards within the TP group
+            assert isinstance(out, torch.Tensor)
+            # TODO: absract out this logic and reuse
+            tensor_local = out
+            tensor_gather = device_mesh.all_gather(tensor_local, mesh_dim=0, gather_dim=0)
+            for other_rank in range(device_mesh.size(dim=0)):
+                if tp_rank != other_rank:
+                    slice_idx = [
+                        slice(other_rank * local_shape[0], (other_rank + 1) * local_shape[0]),
+                        slice(local_shape[1]),
+                    ]
+                    if enable_distribute_flag:
+                        # tensor-parallel dropout results should be different
+                        # within a TP group
+                        self.assertNotEqual(tensor_gather[slice_idx], tensor_local)
+                    else:
+                        # without the parallel RNG, dropout in a tensor-parallel region
+                        # (i.e. on sharded DTensor) will have the same behavior with
+                        # that in a non-tensor-parallel region (i.e. dropout on a 
+                        # replicate DTensor).
+                        self.assertEqual(tensor_gather[slice_idx], tensor_local)
+
+            inp = torch.ones(*local_shape, device=self.device_type)
+            out = resid_dropout_tp(inp)
+            # examine the output
+            # check across TP groups
+            # all-gather local shards across TP groups
+            assert isinstance(out, torch.Tensor)
+            tensor_local = out
+            tensor_gather = device_mesh.all_gather(tensor_local, mesh_dim=1, gather_dim=0)
+            for other_rank in range(device_mesh.size(dim=1)):
```

Comment:
I am wondering if it's possible to make the checking logic be a util function instead? I see both the model init test case and the dropout test case uses similar logic, and we used it for like three times, so we should make this a util instead to avoid redundant code